src/main/java/co/cask/gcp/bigquery/BigQuerySink.java

```diff
@@ -180,116 +114,121 @@ public void prepareRun(BatchSinkContext context) throws Exception {
       BigQueryFileFormat.NEWLINE_DELIMITED_JSON,
       TextOutputFormat.class);
 
-    LineageRecorder lineageRecorder = new LineageRecorder(context, config.referenceName);
-    lineageRecorder.createExternalDataset(config.getSchema());
+    setOutputFormat(context);
+    emitLineage(context, fields);
+  }
 
-    context.addOutput(Output.of(config.referenceName, new OutputFormatProvider() {
-      @Override
-      public String getOutputFormatClassName() {
-        return IndirectBigQueryOutputFormat.class.getName();
-      }
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    super.initialize(context);
+    schema = config.getSchema();
+  }
 
-      @Override
-      public Map<String, String> getOutputFormatConfiguration() {
-        Map<String, String> config = new HashMap<>();
-        for (Map.Entry<String, String> entry : configuration) {
-          config.put(entry.getKey(), entry.getValue());
+  @Override
+  public void transform(StructuredRecord input, Emitter<KeyValue<JsonObject, NullWritable>> emitter) throws Exception {
+    List<Schema.Field> fields = config.getSchema().getFields();
+    JsonObject object = new JsonObject();
+    for (Schema.Field field : fields) {
+      BigQueryUtils.validateSimpleType(field);
```

Comment:
this validation is not required at transform time. It should happen once at configure time